TestRaftRemoveRace: range not found on store 2

The following test appears to have failed:

[#10183](https://circleci.com/gh/cockroachdb/cockroach/10183):

```
W1215 01:18:50.477119 959 multiraft/multiraft.go:1233  aborting configuration change: key range /Local/Range/RangeDescriptor/""-/Min outside of bounds of range /Min-/Min
W1215 01:18:50.478804 959 multiraft/multiraft.go:1139  failed to look up replica ID for range 1 (disabling replica ID check): storage/store.go:1695: store 3 not found as replica of range 1
I1215 01:18:53.557582 959 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
I1215 01:18:53.557889 959 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
I1215 01:18:53.558132 959 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
--- FAIL: TestRaftRemoveRace (3.53s)
	<autogenerated>:32: storage/client_test.go:521: condition failed to evaluate within 3s: storage/client_test.go:517: range not found on store 2
=== RUN   TestStoreRangeRemoveDead
E1215 01:18:53.562437 959 gossip/gossip.go:181  different node IDs were set for the same gossip instance (2147483647, 1)
I1215 01:18:53.563875 959 multiraft/multiraft.go:579  node 1 starting
I1215 01:18:53.564727 959 storage/replica.go:1308  gossiping cluster id  from store 1, range 1
I1215 01:18:53.565694 959 raft/raft.go:446  [group 1] 1 became follower at term 5
I1215 01:18:53.565898 959 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I1215 01:18:53.566074 959 multiraft/multiraft.go:999  node 1 campaigning because initial confstate is [1]
I1215 01:18:53.566196 959 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I1215 01:18:53.566292 959 raft/raft.go:459  [group 1] 1 became candidate at term 6
--
I1215 01:19:05.063638 959 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
I1215 01:19:05.063778 959 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
--- PASS: TestLeaderAfterSplit (0.44s)
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.62s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	32.371s
=== RUN   TestBatchBasics
I1215 01:18:45.307778 1019 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.01s)
=== RUN   TestBatchGet
I1215 01:18:45.312076 1019 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I1215 01:18:45.320845 1019 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.01s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

pretty weird stuff going on here:

```
I1215 01:18:50.464446 959 multiraft/multiraft.go:849  node 3: got message for group 1 with newer replica ID (6 vs 5), recreating group
I1215 01:18:50.466569 959 storage/replica_command.go:451  auto-gc'ed 6ea79b2e (3 intents)
I1215 01:18:50.467203 959 raft/raft.go:446  [group 1] 6 became follower at term 6
I1215 01:18:50.467463 959 raft/raft.go:234  [group 1] newRaft 6 [peers: [], term: 6, commit: 17, applied: 17, lastindex: 17, lastterm: 6]
W1215 01:18:50.467947 959 multiraft/multiraft.go:1139  failed to look up replica ID for range 1 (disabling replica ID check): storage/store.go:1695: store 3 not found as replica of range 1
W1215 01:18:50.473869 959 multiraft/multiraft.go:1139  failed to look up replica ID for range 1 (disabling replica ID check): storage/store.go:1695: store 3 not found as replica of range 1
E1215 01:18:50.475602 959 storage/replica.go:968  error executing raft command: key range /Local/Range/RangeDescriptor/""-/Min outside of bounds of range /Min-/Min
E1215 01:18:50.476069 959 storage/replica.go:968  error executing raft command: key range /Local/Range/RangeDescriptor/""-/Min outside of bounds of range /Min-/Min
W1215 01:18:50.476261 959 multiraft/multiraft.go:1233  aborting configuration change: key range /Local/Range/RangeDescriptor/""-/Min outside of bounds of range /Min-/Min
W1215 01:18:50.476453 959 multiraft/multiraft.go:1233  aborting configuration change: key range /Local/Range/RangeDescriptor/""-/Min outside of bounds of range /Min-/Min
E1215 01:18:50.476918 959 storage/replica.go:968  error executing raft command: key range /Local/Range/RangeDescriptor/""-/Min outside of bounds of range /Min-/Min
W1215 01:18:50.477119 959 multiraft/multiraft.go:1233  aborting configuration change: key range /Local/Range/RangeDescriptor/""-/Min outside of bounds of range /Min-/Min
W1215 01:18:50.478804 959 multiraft/multiraft.go:1139  failed to look up replica ID for range 1 (disabling replica ID check): storage/store.go:1695: store 3 not found as replica of range 1
I1215 01:18:53.557582 959 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
I1215 01:18:53.557889 959 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
I1215 01:18:53.558132 959 storage/engine/rocksdb.go:138  closing in-memory rocksdb instance
	<autogenerated>:32: storage/client_test.go:521: condition failed to evaluate within 3s: storage/client_test.go:517: range not found on store 2
```

--- a/storage/replica_gc_queue.go
+++ b/storage/replica_gc_queue.go
@@ -124,25 +124,7 @@ func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.
 		if log.V(1) {
 			log.Infof("destroying local data from range %d", desc.RangeID)
 		}
-		if err := rng.store.RemoveReplica(rng, replyDesc); err != nil {
-			return err
-		}
-
-		// Lock the store to prevent a new replica of the range from being
-		// added while we're deleting the previous one. We'd really like
-		// to do this before calling RemoveReplica, but this could
-		// deadlock with other work on the Store.processRaft goroutine.
-		// Instead, we check after acquiring the lock to make sure the
-		// range is still absent.
-		q.locker.Lock()
-		defer q.locker.Unlock()
-
-		if _, err := rng.store.getReplicaLocked(desc.RangeID); err == nil {
-			log.Infof("replica recreated during deletion; aborting deletion")
-			return nil
-		}
-
-		if err := rng.Destroy(replyDesc); err != nil {
+		if err := rng.store.RemoveReplica(rng, replyDesc, true); err != nil {
 			return err
 		}
 	} else if desc.RangeID != replyDesc.RangeID {
@@ -153,7 +135,7 @@ func (q *replicaGCQueue) process(now roachpb.Timestamp, rng *Replica, _ *config.
 		if log.V(1) {
 			log.Infof("removing merged range %d", desc.RangeID)
 		}
-		if err := rng.store.RemoveReplica(rng, replyDesc); err != nil {
+		if err := rng.store.RemoveReplica(rng, replyDesc, false); err != nil {
 			return err
 		}
 

--- a/storage/store.go
+++ b/storage/store.go
@@ -1072,7 +1072,7 @@ func (s *Store) MergeRange(subsumingRng *Replica, updatedEndKey roachpb.RKey, su
 
 	// Remove and destroy the subsumed range. Note that we are on the
 	// processRaft goroutine so we can call removeReplicaImpl directly.
-	if err := s.removeReplicaImpl(subsumedRng, *subsumedDesc); err != nil {
+	if err := s.removeReplicaImpl(subsumedRng, *subsumedDesc, false); err != nil {
 		return util.Errorf("cannot remove range %s", err)
 	}
 
@@ -1137,22 +1137,25 @@ func (s *Store) addReplicaToRangeMap(rng *Replica) error {
 type removeReplicaOp struct {
 	rep      *Replica
 	origDesc roachpb.RangeDescriptor
+	destroy  bool
 	ch       chan<- error
 }
 
-// RemoveReplica removes the replica from the store's replica map and from
-// the sorted replicasByKey btree. The version of the replica descriptor that
-// was used to make the removal decision is passed in, and the removal is
-// aborted if the replica ID has changed since then.
-func (s *Store) RemoveReplica(rep *Replica, origDesc roachpb.RangeDescriptor) error {
+// RemoveReplica removes the replica from the store's replica map and
+// from the sorted replicasByKey btree. The version of the replica
+// descriptor that was used to make the removal decision is passed in,
+// and the removal is aborted if the replica ID has changed since
+// then. If `destroy` is true, all data beloing to the replica will be
+// deleted. In either case a tombstone record will be written.
+func (s *Store) RemoveReplica(rep *Replica, origDesc roachpb.RangeDescriptor, destroy bool) error {
 
 	ch := make(chan error)
-	s.removeReplicaChan <- removeReplicaOp{rep, origDesc, ch}
+	s.removeReplicaChan <- removeReplicaOp{rep, origDesc, destroy, ch}
 	return <-ch
 }
 
 // removeReplicaImpl runs on the processRaft goroutine.
-func (s *Store) removeReplicaImpl(rep *Replica, origDesc roachpb.RangeDescriptor) error {
+func (s *Store) removeReplicaImpl(rep *Replica, origDesc roachpb.RangeDescriptor, destroy bool) error {
 	desc := rep.Desc()
 	_, rd := desc.FindReplica(s.StoreID())
 	if rd != nil && rd.ReplicaID >= origDesc.NextReplicaID {
@@ -1167,6 +1170,15 @@ func (s *Store) removeReplicaImpl(rep *Replica, origDesc roachpb.RangeDescriptor
 		return util.Errorf("couldn't find range in replicasByKey btree")
 	}
 	s.scanner.RemoveReplica(rep)
+
+	// TODO(bdarnell): This is fairly expensive to do under store.mu, but doing it outside
+	// the lock is tricky due to the risk that a replica gets recreated by an incoming
+	// raft message.
+	if destroy {
+		if err := rep.Destroy(origDesc); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
@@ -1618,7 +1630,7 @@ func (s *Store) processRaft() {
 			case <-s.wakeRaftLoop:
 
 			case op := <-s.removeReplicaChan:
-				op.ch <- s.removeReplicaImpl(op.rep, op.origDesc)
+				op.ch <- s.removeReplicaImpl(op.rep, op.origDesc, op.destroy)
 
 			case req := <-s.raftRequestChan:
 				if err := s.handleRaftMessage(req); err != nil {

--- a/storage/store_test.go
+++ b/storage/store_test.go
@@ -281,7 +281,7 @@ func TestStoreAddRemoveRanges(t *testing.T) {
 		t.Error(err)
 	}
 	// Remove range 1.
-	if err := store.RemoveReplica(rng1, *rng1.Desc()); err != nil {
+	if err := store.RemoveReplica(rng1, *rng1.Desc(), true); err != nil {
 		t.Error(err)
 	}
 	// Create a new range (id=2).
@@ -298,7 +298,7 @@ func TestStoreAddRemoveRanges(t *testing.T) {
 		t.Fatalf("expected rangeAlreadyExists error; got %s", err)
 	}
 	// Try to remove range 1 again.
-	if err := store.RemoveReplica(rng1, *rng1.Desc()); err == nil {
+	if err := store.RemoveReplica(rng1, *rng1.Desc(), true); err == nil {
 		t.Fatal("expected error re-removing same range")
 	}
 	// Try to add a range with previously-used (but now removed) ID.
@@ -354,12 +354,12 @@ func TestStoreRemoveReplicaOldDescriptor(t *testing.T) {
 	if err := rng1.setDesc(newDesc); err != nil {
 		t.Fatal(err)
 	}
-	if err := store.RemoveReplica(rng1, *origDesc); !testutils.IsError(err, "replica ID has changed") {
+	if err := store.RemoveReplica(rng1, *origDesc, true); !testutils.IsError(err, "replica ID has changed") {
 		t.Fatalf("expected error 'replica ID has changed' but got %s", err)
 	}
 
 	// Now try the latest descriptor and succeed.
-	if err := store.RemoveReplica(rng1, *rng1.Desc()); err != nil {
+	if err := store.RemoveReplica(rng1, *rng1.Desc(), true); err != nil {
 		t.Fatal(err)
 	}
 }
@@ -374,7 +374,7 @@ func TestStoreRangeSet(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	if err := store.RemoveReplica(rng1, *rng1.Desc()); err != nil {
+	if err := store.RemoveReplica(rng1, *rng1.Desc(), true); err != nil {
 		t.Error(err)
 	}
 	// Add 10 new ranges.
@@ -457,7 +457,7 @@ func TestStoreRangeSet(t *testing.T) {
 	if rng.RangeID != 2 {
 		t.Errorf("expected fetch of rangeID=2; got %d", rng.RangeID)
 	}
-	if err := store.RemoveReplica(rng, *rng.Desc()); err != nil {
+	if err := store.RemoveReplica(rng, *rng.Desc(), true); err != nil {
 		t.Error(err)
 	}
 	if ec := ranges.EstimatedCount(); ec != 9 {
@@ -481,7 +481,7 @@ func TestHasOverlappingReplica(t *testing.T) {
 		t.Error(err)
 	}
 	// Remove range 1.
-	if err := store.RemoveReplica(rng1, *rng1.Desc()); err != nil {
+	if err := store.RemoveReplica(rng1, *rng1.Desc(), true); err != nil {
 		t.Error(err)
 	}
 
@@ -1641,7 +1641,7 @@ func TestMaybeRemove(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
-	if err := store.RemoveReplica(rng, *rng.Desc()); err != nil {
+	if err := store.RemoveReplica(rng, *rng.Desc(), true); err != nil {
 		t.Error(err)
 	}
 	// MaybeRemove is called.

